Userland/Libraries/LibThreading/BackgroundAction.h

```diff
@@ -34,51 +35,71 @@ class BackgroundActionBase {
 
-public:
-    void cancel()
+    ResultWrapper(Result& result)
+        : result(result)
     {
-        m_cancelled = true;
     }
 
-    bool is_cancelled() const
-    {
-        return m_cancelled;
-    }
+    Result& result;
+};
+
+// BackgroundAction has to inherit from Promise<Object> to allow it to be managed by EventLoop.
+// However, the internal result type is always ResultWrapper<Result>.
+template<typename Result>
+class BackgroundAction final : public Core::Promise<NonnullRefPtr<Core::Object>>
+    , private BackgroundActionBase {
+    C_OBJECT(BackgroundAction);
+
+public:
+    using Promise = Core::Promise<NonnullRefPtr<Core::Object>>;
 
     virtual ~BackgroundAction() = default;
 
+    Optional<Result> const& result() const { return m_result; }
+
 private:
-    BackgroundAction(Function<Result(BackgroundAction&)> action, Function<ErrorOr<void>(Result)> on_complete, Optional<Function<void(Error)>> on_error = {})
-        : Core::Object(&background_thread())
+    BackgroundAction(Function<Result(BackgroundAction&)> action, Function<ErrorOr<void>(Result&)> on_resolved, Optional<Function<void(Error)>> on_error = {})
+        : Promise(&background_thread())
         , m_action(move(action))
-        , m_on_complete(move(on_complete))
     {
+        if (on_resolved) {
+            this->on_resolved = [on_resolved = move(on_resolved)](auto& wrapped_result) {
+                return on_resolved(static_ptr_cast<ResultWrapper<Result>>(wrapped_result)->result);
+            };
+        }
+
         if (on_error.has_value())
             m_on_error = on_error.release_value();
 
+        if (this->on_resolved)
+            Core::EventLoop::current().add_job(*this);
+
         enqueue_work([this, origin_event_loop = &Core::EventLoop::current()] {
-            m_result = m_action(*this);
-            if (m_on_complete) {
-                origin_event_loop->deferred_invoke([this] {
-                    auto maybe_error = m_on_complete(m_result.release_value());
-                    if (maybe_error.is_error())
-                        m_on_error(maybe_error.release_error());
-                    remove_from_parent();
-                });
-                origin_event_loop->wake();
+            this->m_result = m_action(*this);
+            this->m_pending = ResultWrapper<Result>::construct(m_result.value());
+            if (this->on_resolved) {
+                if (!this->is_cancelled()) {
+                    origin_event_loop->deferred_invoke([this] {
+                        auto maybe_error = this->on_resolved(this->m_pending.value());
+                        if (maybe_error.is_error())
+                            this->m_on_error(maybe_error.release_error());
+                        this->remove_from_parent();
+                    });
+                    origin_event_loop->wake();
+                } else {
+                    dbgln("BackgroundAction: Ran longer than the event loop, not invoking the on_resolved callback.");
```

Comment:
This debug statement is misleading: At this point we only know that the BackgroundAction has been canceled, but not necessarily why. Maybe adjust it slightly?